### ISSUE TITLE
[Suggestion] - Library chart which contains common logic between Atlassian charts

### ISSUE THREAD

**Issue by hickeyma** [atlassian/data-center-helm-charts#302]
### Suggestion

There is a duplication of logic between charts. An examples of this  is the `fullname` definition which is duplicated in each of the charts. For example `bitbucket` and `jira` examples below:

```yaml
{{/*
Create a default fully qualified app name.
We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
If release name contains chart name it will be used as a full name.
*/}}
{{- define "bitbucket.fullname" -}}
{{- if .Values.fullnameOverride }}
{{- .Values.fullnameOverride | trunc 63 | trimSuffix "-" }}
{{- else }}
{{- $name := default .Chart.Name .Values.nameOverride }}
{{- if contains $name .Release.Name }}
{{- .Release.Name | trunc 63 | trimSuffix "-" }}
{{- else }}
{{- printf "%s-%s" .Release.Name $name | trunc 63 | trimSuffix "-" }}
{{- end }}
{{- end }}
{{- end }}
```

```yaml
{{/*
Create a default fully qualified app name.
We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
If release name contains chart name it will be used as a full name.
*/}}
{{- define "jira.fullname" -}}
{{- if .Values.fullnameOverride }}
{{- .Values.fullnameOverride | trunc 63 | trimSuffix "-" }}
{{- else }}
{{- $name := default .Chart.Name .Values.nameOverride }}
{{- if contains $name .Release.Name }}
{{- .Release.Name | trunc 63 | trimSuffix "-" }}
{{- else }}
{{- printf "%s-%s" .Release.Name $name | trunc 63 | trimSuffix "-" }}
{{- end }}
{{- end }}
{{- end }}
```

These definitions are essentially the same logic which could be defined in one chart, a [Helm Library Chart](https://helm.sh/docs/topics/library_charts/#helm) and used between the Atlassian charts.

Some of the benefits of introducing a common or library chart are:

- Code re-use and keeping charts [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- Minimizing errors
- Maintenance improvements

**Note:** The common logic can be extracted from the charts and added to the common chart on a bit by bit basis and avoid "big bang" approach. 

### Product

Jira, Confluence, Bitbucket, Other

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by hickeyma** [comment]
Pushed PR #303 which creates the library (common) chart and adds some of the common logic definitions. This also provides a PoC of the feature.

**Comment by nanux** [comment]
Hi @hickeyma, this is an interesting change that we have considered as well. Thank you very much for the PR it is something we can test and see how to work with the current installations and going forward. We haven't introduced an "umbrella" chart as we didn't want to have explicit product dependency on other charts but with more duplication, this should make it easier for maintaining the helm charts and also maintaining installations.

Right now, there is another major work currently in progress that might utilise this but for now, we would like to avoid any impact on the development so I will put the issue on hold and revisit it within a ~month.

**Comment by hickeyma** [comment]
@nanux That's great that you have considered it as it is cuts down on code repetition as its core benefit. For clarification, this is not an "umbrella" chart, it is a "common" or "library" chart which can be used by other charts. An "umbrella" chart composes the different parts of a complex application as one top level chart. See [Complex Charts with Many Dependencies](https://helm.sh/docs/howto/charts_tips_and_tricks/#complex-charts-with-many-dependencies) for more details.

> Right now, there is another major work currently in progress that might utilise this but for now, we would like to avoid any impact on the development so I will put the issue on hold and revisit it within a ~month.

Not a bother. Once you have completed this work, let me know and we can see about adding the functionality.

**Comment by hickeyma** [comment]
Hey @nanux, any update?

**Comment by nanux** [comment]
Hi @hickeyma, sorry it took some time, there were other moving pieces that I was working on. I think we are now at the point where we can start introducing the library chart. Do you think you can pick up the PR and update it to make it work with the latest changes?
Thank you

**Comment by hickeyma** [comment]
@nanux Sure, will rebase it and push the update.

**Comment by hickeyma** [comment]
@nanux I have updated PR #303, its waiting on approval to run the workflow.
